polkadot/xcm/pallet-xcm/src/tests.rs

```diff
@@ -433,186 +560,1570 @@ fn limited_teleport_assets_works() {
+		let dest = usdc_reserve_location;
+
+		let (assets, fee_index, fee_asset, xfer_asset) = into_multiassets_checked(
+			// usdc for fees (is sufficient on local chain too) - destination reserve
+			(usdc_id_multilocation, FEE_AMOUNT).into(),
+			// native asset to transfer (not used for fees) - local reserve
+			(MultiLocation::here(), SEND_AMOUNT).into(),
+		);
+
+		// reanchor according to test-case
+		let context = UniversalLocation::get();
+		let expected_fee = fee_asset.reanchored(&dest, context).unwrap();
+		let expected_asset = xfer_asset.reanchored(&dest, context).unwrap();
+
+		// balances checks before
+		assert_eq!(Assets::balance(usdc_id_multilocation, ALICE), usdc_initial_local_amount);
+		assert_eq!(Balances::free_balance(ALICE), INITIAL_BALANCE);
+
+		// do the transfer
+		assert_ok!(XcmPallet::limited_reserve_transfer_assets(
 			RuntimeOrigin::signed(ALICE),
-			Box::new(Parachain(PARA_ID).into()),
 			Box::new(dest.into()),
-			Box::new((Here, SEND_AMOUNT).into()),
-			0,
+			Box::new(beneficiary.into()),
+			Box::new(assets.into()),
+			fee_index as u32,
+			Unlimited,
 		));
-		// Alice spent amount
+		assert!(matches!(
+			last_event(),
+			RuntimeEvent::XcmPallet(crate::Event::Attempted { outcome: Outcome::Complete(_) })
+		));
+
+		// Alice spent (fees) amount
+		assert_eq!(
+			Assets::balance(usdc_id_multilocation, ALICE),
+			usdc_initial_local_amount - FEE_AMOUNT
+		);
+		// Alice used native asset for transfer
 		assert_eq!(Balances::free_balance(ALICE), INITIAL_BALANCE - SEND_AMOUNT);
-		// Destination account (parachain account) has amount
-		let para_acc: AccountId = ParaId::from(PARA_ID).into_account_truncating();
-		assert_eq!(Balances::free_balance(para_acc), INITIAL_BALANCE + SEND_AMOUNT);
+		// Sovereign account of dest parachain holds `SEND_AMOUNT` native asset in local reserve
+		assert_eq!(Balances::free_balance(usdc_chain_sovereign_account.clone()), SEND_AMOUNT);
+		assert_eq!(Assets::balance(usdc_id_multilocation, usdc_chain_sovereign_account), 0);
+		// Verify total and active issuance of USDC have decreased (burned on reserve-withdraw)
+		let expected_issuance = usdc_initial_local_amount - FEE_AMOUNT;
+		assert_eq!(Assets::total_issuance(usdc_id_multilocation), expected_issuance);
+		assert_eq!(Assets::active_issuance(usdc_id_multilocation), expected_issuance);
+
+		// Verify sent XCM program
+		assert_eq!(
+			sent_xcm(),
+			vec![
+				(
+					// first message is to prefund fees on `dest`
+					dest,
+					// fees are being sent through destination-reserve transfer because fee reserve
+					// is destination chain
+					Xcm(vec![
+						WithdrawAsset(expected_fee.clone().into()),
+						ClearOrigin,
+						buy_limited_execution(expected_fee.clone(), Unlimited),
+						DepositAsset { assets: AllCounted(1).into(), beneficiary },
+					])
+				),
+				(
+					// second message is to transfer/deposit (native) asset on `dest` while paying
+					// using prefunded (transferred above) fees
```

Comment:
How do you pay using prefunded fees?
Unless fees mode is set to jit_withdraw, you'd need to withdraw the fees from the account on the second message

Reply:
yes, good catch!

I had the same realization recently and was looking at how we can add tests to xcm-emulator for these scenarios to validate if they actually work as expected

Reply:
went with `SetFessMode { jit_withdraw: <case-by-case> }` approach as it should be cleanest/most extensible/flexible